### PR TITLE
Fix C/C++ debug hover on intermediate members of a dereferenced expression

### PR DESCRIPTION
**Pull request by tieo**
## Problem

While debugging C/C++, hovering an intermediate member of a `*`/`&`-prefixed access chain (e.g. `a` or `b` in `if (*a.b.c == X)` where `a.b` is a struct) shows no value, and hovering a member that comes after an array subscript (e.g. `c` in `a.b[i].c`) shows no value either.

## Cause

No `EvaluatableExpressionProvider` is registered for C/C++, so VS Code's built-in data-tip expression detection is used. It keeps the leading `*`/`&` and clips on the right, so hovering `b` in `*a.b.c` evaluates `*a.b` — a dereference of the non-pointer member `a.b` — which errors. It also stops at `[` and `]`, so hovering `c` in `a.b[i].c` evaluates a broken fragment after the `]`.

## Fix

Register an `EvaluatableExpressionProvider` for C/C++. Once a provider is registered, VS Code uses it instead of its built-in detection (it does not fall back to the default when the provider returns nothing), so the provider reproduces the built-in behavior for ordinary tokens and additionally fixes the access-chain cases:

- A leading `*`/`&` applies to the whole access chain (the postfix `.`, `->`, `[]` operators bind tighter). It is dropped for an interior member of a `.` chain, where `*a.b` would dereference the struct `a.b`, and kept on the final member and before `->` (so `*ptr->member` is unchanged).
- Array subscripts and `::` are kept in the token, so members after a subscript (`a.b[i].c`) and scoped names (`ns::var`) resolve. Hovering a subscript bracket evaluates the indexed element; hovering inside `[...]` evaluates the index on its own.

The provider is registered during debugger activation (so it also works when cpptools IntelliSense is disabled), and the expression computation lives in a `vscode`-free module (`evaluatableExpression.ts`) with unit tests.

Note: I created the code in this PR with the help of an LLM, and tested it manually in VS Code plus with the added unit tests.

### CODE REVIEW

**Comment by tieo** [comment]
@microsoft-github-policy-service agree

File: Extension/src/LanguageServer/client.ts [hunk 1]
```diff
@@ -1431,6 +1432,7 @@ export class DefaultClient implements Client {
 
                 this.disposables.push(vscode.languages.registerHoverProvider(util.documentSelector, instrument(this.copilotHoverProvider)));
                 this.disposables.push(vscode.languages.registerHoverProvider(util.documentSelector, instrument(this.hoverProvider)));
+                this.disposables.push(vscode.languages.registerEvaluatableExpressionProvider(util.documentSelector, instrument(new EvaluatableExpressionProvider())));
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/client.ts]: I think this belongs in vscode-cpptools\Extension\src\Debugger\extension.ts

File: Extension/src/LanguageServer/Providers/evaluatableExpressionProvider.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/Providers/evaluatableExpressionProvider.ts]: I think this belongs under vscode-cpptools\Extension\src\Debugger.  vscode-cpptools\Extension\src\LanguageServer is for the non-debugger IntelliSense features.

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]
...see my other comments.

File: Extension/src/LanguageServer/Providers/evaluatableExpressionProvider.ts (hunk 2 above)
Comment (sean-mcmanus) [reply Extension/src/LanguageServer/Providers/evaluatableExpressionProvider.ts]: ...just for organization and not a technical API reason, since we split up the debugger and language server features separately.

**Comment by tieo** [comment]
Addressed the review feedback in the latest commit:

- Moved the provider to `Extension/src/Debugger/` and register it during debugger activation in `Debugger/extension.ts` instead of the language client.

On the Copilot suggestion to gate with `line.charAt(clipEnd) === '.'`: I went a different way. The problem with that gate is it leaves `*ptr->member` doing what the default does, but that's already broken — `->` has higher precedence than the unary `*`, so the expression is really `*(ptr->member)`, and when you hover `ptr` you want to see `ptr`, not `*ptr`.

What it does now is keep the leading `*`/`&` only when you're hovering the last segment of the chain, and drop it everywhere in the middle. That ends up being right whether the chain uses `.`, `->`, or `[]`. I also had to keep subscripts in the token because the default detection stops at the `[`, which is why hovering something like the `c` in `a.b[i].c` gave you nothing before. Function calls never get pulled in, so there's no risk of evaluating something with side effects.

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]
My local Copilot is saying the following:

**The implementation doesn't match the description — the central claim is inverted.**

The PR says:

> Every other case returns `undefined`, so the default behavior (`.`/`->`/`::`, and legitimate `*ptr` derefs on the final segment) is unchanged.

But the code never returns `undefined` except when no token is found under the cursor. For essentially every token it computes and returns an expression. Its own header comment states the opposite of the PR:

> While a provider is registered VS Code does not apply its built-in expression detection, so an expression is returned for any token rather than `undefined`.

The API contract disagrees with that comment — `provideEvaluatableExpression` documents that `undefined`/`null` signals "lack of a result," and the editor falls back to default word detection. So the surgical design the PR describes is exactly what the API supports, but the code implements a **full takeover** of C/C++ hover-expression detection instead. Please reconcile: either make the code actually return `undefined` outside the one target case (matches the description, lowest risk), or rewrite the description and risk assessment to state that this provider now owns *all* C/C++ data-tip expressions. The first option is strongly preferable.

**Consequences of the takeover that the description says are safe but aren't:**

1. **`::` is silently broken.** The description lists `::` as preserved by default behavior, but the token grammar (`(?:[\p{L}\p{N}_]+|->|\.|\[[^\][]*\])+`) has no `::`. So `Foo::bar` gets clipped to `bar`/`Foo` and the qualified expression the default would have produced is overridden. Returning `undefined` for this case makes it moot; keeping the takeover means `::` must be added to the grammar.

2. **Numeric/other tokens now resolve to themselves** (`42`, `1.5`) because the provider answers for them too. Again moot under the surgical approach.

**The fix is also incomplete relative to its own bug class.** The Problem uses `*a.b.c`, but the same intermediate-dereference bug exists for `->` and `[]`, and the code doesn't handle them. The gate is `line.charAt(clipEnd) !== '.'`, so:

- `*a->b`, hovering `a` → still `*a` (precedence makes this `*(a->b)`; should be `a`)
- `*a.b[i]`, hovering `b` → still `*a.b`

This contradicts your earlier note that it should "drop everywhere in the middle… whether the chain uses `.`, `->`, or `[]`." Drop the `.`-only check so any intermediate segment strips the leading operator.

**Contradictory comments.** The header says the operator is dropped "on an interior member or a subscript"; the inline comment says it's kept "before `->` or a subscript." Both can't hold, and neither matches the last-segment-only rule. Make them consistent with the final behavior.

**Recommendation:** implement it the way the description already promises — compute the stripped expression *only* for an intermediate member of a `*`/`&` chain (covering `.`, `->`, and `[]`), and return `undefined` in every other case. That fixes the `::`/numeric regressions, completes the `->`/`[]` coverage, restores the "default behavior unchanged" guarantee, and makes the code match the PR.

**Comment by tieo** [comment]
Thanks — addressed the feedback and rewrote the description, since the old one no longer matched the code.

On the "returns `undefined` otherwise" point: that was the original intent, but it doesn't actually work. Once an `EvaluatableExpressionProvider` is registered, VS Code uses it *instead of* its built-in detection and does not fall back to the default when the provider returns `undefined` — the data-tip just shows nothing. I verified this by running the resolver's own logic: with a provider registered that returns `undefined`, it returns `null` (no tip); the built-in extraction only runs when no provider is registered at all. So the provider has to return a sensible expression for ordinary tokens, not `undefined`. I updated the description and comments to say this.

On `::`: added to the tokenizer, so `ns::var` stays together.

On `<=` vs `<` in the hit-test: I kept `<=`. VS Code's own `getExactExpressionStartAndEnd` uses an inclusive end (`l >= e`), which maps to `<=` here — so `<=` matches the built-in detection, and `<` would diverge by failing to select a token when hovering its trailing edge.

On tests: moved the computation into a `vscode`-free module (`evaluatableExpression.ts`) and added unit tests covering `*a.b.c` interior vs final, `->`, `::`, subscripts (`a.b[i].c`, bracket vs index), and the no-token case.

File: Extension/src/Debugger/evaluatableExpressionProvider.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        if (openCount > closeCount) {
+            return new vscode.EvaluatableExpression(new vscode.Range(position.line, wordStart, position.line, clipEnd), word);
+        }
+
+        // The access chain from its start through the identifier under the cursor.
+        const defaultExpression = (): vscode.EvaluatableExpression =>
+            new vscode.EvaluatableExpression(new vscode.Range(position.line, tokenStart, position.line, clipEnd), line.substring(tokenStart, clipEnd));
+
+        if (leading === null) {
+            return defaultExpression();
```
[the remark this answers is not in the document]
Comment (sean-mcmanus) [reply Extension/src/Debugger/evaluatableExpressionProvider.ts]: Seems like this isn't 100% resolved yet: 

**But the actual concern the comment raised — "can diverge from VS Code's built-in detection" — is still live.** There's no parity test, and the divergence is real: tokens can begin with a connector, so `foo().bar` / `(*this).member` / `obj->fn()->field` return broken fragments (`.bar`, `.member`, `->field`). So the intent is addressed and documented, but the divergence it warned about is not eliminated.

This is the same as the two new issues raised by the GitHub Copilot.

File: Extension/test/unit/evaluatableExpression.test.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ * ------------------------------------------------------------------------------------------ */
+
+import { describe, it } from 'mocha';
+import { strictEqual } from 'node:assert';
+import { computeEvaluatableExpression } from '../../src/Debugger/evaluatableExpression';
+
+// In each input the `|` marks the cursor; it is removed before evaluating.
+function evaluate(marked: string): string | undefined {
+    const character: number = marked.indexOf('|');
+    const line: string = marked.replace('|', '');
```
[the remark this answers is not in the document]
Comment (sean-mcmanus) [reply Extension/test/unit/evaluatableExpression.test.ts]: ## CodeQL `js/incomplete-sanitization` (test helper)

It's a benign warning on test code, but trivially silenced and made more robust by slicing at the known marker index instead of `String.replace`, which also documents the single-marker assumption:

```ts
function evaluate(marked: string): string | undefined {
    const character: number = marked.indexOf('|');
    const line: string = marked.slice(0, character) + marked.slice(character + 1);
    return computeEvaluatableExpression(line, character)?.expression;
}
```

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]
My local Copilot says the following (the 1st issue is the same as the 2 comments made by the GitHub Copilot).

## Significant: tokens can begin with a connector, producing broken expressions

The token grammar lets a match start with `.`, `->`, or `::`, because the repeated group's alternatives are tried independently:

```
/(?:[*&]+)?(?:[\p{L}\p{N}_]+|->|::|\.|\[[^\][]*\])+/gu
```

For any expression where a primary isn't a bare identifier — parenthesized or call results — the regex skips the un-matchable `(`/`)` and starts a fresh token at the connector. Tracing `foo().bar`:

- `foo` matches, stops at `(`
- `(` and `)` are unmatchable, scanned past
- match resumes at `.` → token **`.bar`**

Hovering `bar` returns `.bar`, which gets sent to the debugger and errors. Same for:

- `(*this).member` → hovering `member` → `.member`
- `obj->method()->field` → hovering `field` → `->field`

These are common patterns, and built-in detection likely produced `bar`/`member`/`field` for them, so this is a regression risk, not just an edge case. Fix: anchor the chain so it must start with an identifier (or leading op + identifier), e.g.

```js
/(?:[*&]+)?[\p{L}\p{N}_]+(?:[\p{L}\p{N}_]+|->|::|\.|\[[^\][]*\])*/gu
```

and add tests for `foo().bar`, `(*this).member`, and `obj->fn()->field` so a leading-connector expression can never be returned. (Optional, while you're there: tightening the identifier start to `[\p{L}_][\p{L}\p{N}_]*` would also stop `1.5`/`0x1F` numeric literals from being tokenized as member chains — a smaller parity divergence.)

## Confirm intent: `*ptr->member` hovering `ptr` → `*ptr`

The new test locks in `evaluate('*|ptr->member') === '*ptr'`. Worth flagging because in an earlier note you argued the opposite — "when you hover `ptr` you want to see `ptr`, not `*ptr`." The current behavior is actually defensible: `ptr->` proves `ptr` is a pointer, so `*ptr` is a valid dereference (it won't error the way `*a.b` does), and it matches built-in's keep-leading-and-clip-right behavior. But it does show the pointee rather than the pointer the user pointed at. The same family of cases is also now locked in by the `.`-only drop:

- `*a.b[i]` hovering `b` → `*a.b` (valid — array decays — but shows the first element)
- `&a[i]` hovering `a` → `&a` (address of the array)

None of these error, so they don't reintroduce the "shows nothing" bug, but they're semantically surprising. If `*ptr`/`*a.b`/`&a` are the intended answers, fine — just make it a deliberate, documented choice rather than a side effect of the `charAt(clipEnd) !== '.'` gate. A one-line comment on why `.` is special (it's the only connector whose left operand isn't provably a pointer, so it's the only one where keeping `*` would error) would make this self-explanatory.

## Summary

- **Must fix:** anchor the token so it can't start with `.`/`->`/`::` (avoids returning `.bar`/`->field` for `foo().bar`, `(*this).member`, etc.); add tests for those.
- **Confirm + comment:** the `*ptr`/`*a.b`/`&a` keep-leading outcomes are intentional (they don't error, but they're surprising) — add a one-line rationale for the `.`-only drop.

**Comment by tieo** [comment]
Thanks for the detailed review. I've addressed the remaining points.

**Connector-leading fragments.** The provider now declines these explicitly: after any leading `*`/`&`, if the token begins with `.` or `->`, it returns `undefined`. So `foo().bar`, `(*this).member`, and `obj->fn()->field` no longer produce `.bar`, `.member`, or `->field`. I added regression tests for all three. I guard at the start of the token rather than anchoring the regex to an identifier, because that also keeps a legitimate leading `::` global-scope qualifier (e.g. `::var`) working, which an identifier anchor would reject.

**The `.`-only drop of the leading operator.** This is intentional, and I've added a comment explaining why. `.` is the only connector whose left operand is not provably a pointer or array, so keeping the leading operator there (`*a.b`) would dereference a struct and error. Before `->` or `[]`, and at the end of the chain, the left operand is a pointer or array, so the leading operator stays valid. That makes `*ptr->member`, and deliberately `*a.b[i]` and `&a[i]`, evaluate without error (the array decays); they show the element or address rather than the hovered operand, which matches the built-in keep-leading-and-clip-right behavior.

**CodeQL.** The test helper now slices at the marker index instead of using `String.replace`, and I tightened the tokenizer to avoid a nested-quantifier pattern that the ReDoS query flags.

On the optional suggestion to require an identifier start so numeric literals like `1.5` or `0x1F` aren't tokenized as member chains: I left this as-is, since those evaluate without error in a data-tip and match the built-in detection. Happy to add it if you'd prefer the stricter parity.

File: Extension/src/Debugger/evaluatableExpression.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    // An identifier inside a `[...]` is the index; it is evaluated on its own, not as part of the
+    // surrounding access chain.
+    const beforeCursor: string = line.substring(tokenStart, character);
+    const openCount: number = (beforeCursor.match(/\[/g) || []).length;
+    const closeCount: number = (beforeCursor.match(/\]/g) || []).length;
+    if (openCount > closeCount) {
+        return { startColumn: wordStart, endColumn: clipEnd, expression: word };
+    }
```
[the remark this answers is not in the document]
Comment (tieo) [reply Extension/src/Debugger/evaluatableExpression.ts]: Good catch — fixed in 618964a4. Inside a subscript the token spans whitespace and operators, so the index identifier is now only returned when the cursor is actually on it; otherwise the position is treated as not on a token and returns `undefined`. Added tests for `a[i + j]` with the cursor on the operator and the surrounding spaces.

**Comment by tieo** [comment]
Pushed a follow-up that reworks the token detection after testing against real debug sessions. Summary of what changed and why:

**Balanced-bracket scanner instead of a regex.** The token is now found by a small manual scanner rather than `tokenRegExp`. This fixes nested subscripts (`a[b[i]]`, which the regex split at the inner `[`), and it removes the previous regex entirely — so the earlier `js/redos` concern is moot.

**Leading-operator rule, corrected.** My earlier comment described `*a.b[i]` / `&a[i]` keeping the leading operator as "deliberate." Testing in a real debugger showed that was wrong: `*a.b[i]` dereferences the array base (element 0) and `&a[i]` shows an address, neither of which is what you hovered. The rule is now:
- A leading `*` is kept only on the **final dereferenced segment** — `*a.b.c` (final member), and the final subscript element `*a.b[i]` (the dereferenced element you're pointing at).
- It's dropped on any **interior** segment: hovering `b` in `*a.b.c` → `a.b`, and `b` in `*a.b[i]` → `a.b` (the array, not `*a.b`).
- A leading `&` is **always dropped**, so hovering the variable shows its value, not its address.

**Subscript edge cases.** Non-token positions inside `[...]` (operators/whitespace, e.g. `a[i + j]`) return undefined; hovering past the last identifier no longer truncates a trailing subscript.

Unit tests cover all of the above (interior vs final `*`, `&`, nested subscripts, the index, connector-leading fragments, and the operator/whitespace cases).

Separately: while testing I hit a data-tip rendering issue where a `uint32_t` scalar shows as an address (the memory icon reads the value as an address). That turned out to be in the debug adapter's `memoryReference` handling, not expression detection — cpptools' TypeScript never sets `memoryReference`, so it's out of scope for this PR. I'll raise it separately.

**Review by sean-mcmanus — APPROVED** [review]

**Review by sean-mcmanus — APPROVED** [review]
Seems good to me. I'm not sure if anyone else on our team wants to review it still. I think it can get into our next 1.33.3 release...not sure when that will be yet though.

**Comment by sean-mcmanus** [comment]
@tieo Your fix is available in https://github.com/microsoft/vscode-cpptools/releases/tag/v1.33.3 pre-release.
